From 9ec5a0a776b539deb44f34fc6973969c4b7dbda4 Mon Sep 17 00:00:00 2001
From: Jan Kiszka <jan.kiszka@siemens.com>
Date: Fri, 30 Apr 2021 07:57:59 +0200
Subject: [PATCH 10/13] uart: Fix software flow control management

Rather than updating IXON/IXOFF in termios, mraa_uart_set_flowcontrol
was incorrectly issuing a stop or start character on mode changes. This
lead to spurious transmission in setups that actually wanted to disable
software flow control. And it prevented enabling it (which could have
been checked also by reading back the state via mraa_uart_settings).

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
---
 src/uart/uart.c | 16 ++++++----------
 1 file changed, 6 insertions(+), 10 deletions(-)

diff --git a/src/uart/uart.c b/src/uart/uart.c
index f1405f8..3e3aa32 100644
--- a/src/uart/uart.c
+++ b/src/uart/uart.c
@@ -641,16 +641,6 @@ mraa_uart_set_flowcontrol(mraa_uart_context dev, mraa_boolean_t xonxoff, mraa_bo
         }
     }
 
-    // hardware flow control
-    int action = TCIOFF;
-    if (xonxoff) {
-        action = TCION;
-    }
-    if (tcflow(dev->fd, action)) {
-        return MRAA_ERROR_FEATURE_NOT_SUPPORTED;
-    }
-
-    // rtscts
     struct termios termio;
 
     // get current modes
@@ -659,6 +649,12 @@ mraa_uart_set_flowcontrol(mraa_uart_context dev, mraa_boolean_t xonxoff, mraa_bo
          return MRAA_ERROR_INVALID_RESOURCE;
     }
 
+    if (xonxoff) {
+        termio.c_iflag |= IXON|IXOFF;
+    } else {
+        termio.c_iflag &= ~(IXON|IXOFF);
+    }
+
     if (rtscts) {
         termio.c_cflag |= CRTSCTS;
     } else {
-- 
2.26.2

